pip/src/noisy_simulator.rs

```diff
@@ -0,0 +1,320 @@
+        }
+    }
+    list
+}
+
+pyo3::create_exception!(qsharp.noisy_sim, SimulationError, PyException);
+
+#[pyclass]
+#[derive(Clone)]
+pub(crate) struct Operation(noisy_simulator::Operation);
+
+#[pymethods]
+impl Operation {
+    #[new]
+    pub fn new(kraus_operators: Vec<PythonMatrix>) -> Self {
+        let kraus_operators: Vec<SquareMatrix> = kraus_operators
+            .into_iter()
+            .map(python_to_nalgebra_matrix)
+            .collect();
+        Self(noisy_simulator::Operation::new(kraus_operators))
+    }
+
+    pub fn get_effect_matrix(&self) -> Vec<Complex<f64>> {
+        nalgebra_matrix_to_python_list(self.0.effect_matrix())
+    }
+
+    pub fn get_operation_matrix(&self) -> Vec<Complex<f64>> {
+        nalgebra_matrix_to_python_list(self.0.matrix())
+    }
+
+    pub fn get_kraus_operators(&self) -> Vec<Vec<Complex<f64>>> {
+        let mut kraus_operators = Vec::new();
+        for kraus_operator in self.0.kraus_operators() {
+            kraus_operators.push(nalgebra_matrix_to_python_list(kraus_operator));
+        }
+        kraus_operators
+    }
+}
+
+#[pyclass]
+pub(crate) struct Instrument(noisy_simulator::Instrument);
+
+#[pymethods]
+impl Instrument {
+    #[new]
+    pub fn new(operations: Vec<Operation>) -> Self {
+        let operations = operations.into_iter().map(|op| op.0).collect();
+        Self(noisy_simulator::Instrument::new(operations))
+    }
+}
+
+#[pyclass]
+#[derive(Clone)]
+pub(crate) struct DensityMatrix {
+    /// Dimension of the matrix. E.g.: If the matrix is 5 x 5, then dim is 5.
+    dim: usize,
+    /// Number of qubits in the system.
+    number_of_qubits: usize,
+    /// Theoretical change in trace due to operations that have been applied so far.
+    trace_change: f64,
+    /// Vector storing the entries of the density matrix.
+    data: Vec<Complex<f64>>,
+}
+
+impl From<&noisy_simulator::DensityMatrix> for DensityMatrix {
+    fn from(dm: &noisy_simulator::DensityMatrix) -> Self {
+        Self {
+            dim: dm.dim(),
+            number_of_qubits: dm.number_of_qubits(),
+            trace_change: dm.trace_change(),
+            data: dm.data().iter().copied().collect::<Vec<_>>(),
+        }
+    }
+}
+
+impl TryInto<noisy_simulator::DensityMatrix> for DensityMatrix {
+    type Error = PyErr;
+
+    fn try_into(self) -> PyResult<noisy_simulator::DensityMatrix> {
+        noisy_simulator::DensityMatrix::try_from(
+            self.dim,
+            self.number_of_qubits,
+            self.trace_change,
+            ComplexVector::from_vec(self.data),
+        )
+        .ok_or(SimulationError::new_err("invalid density matrix"))
+    }
+}
+
+#[pyclass]
+pub(crate) struct DensityMatrixSimulator(noisy_simulator::DensityMatrixSimulator);
+
+#[pymethods]
+impl DensityMatrixSimulator {
+    #[new]
+    #[pyo3(signature = (number_of_qubits, seed=42))]
+    #[allow(unused_variables)]
+    pub fn new(number_of_qubits: usize, seed: usize) -> Self {
```

Comment:
remove default value, make seen an option, where is seed consumed?

Reply:
This was to be able to use this package as a drop-in replacement for the Cpp simulator. Since we don't have that requirement anymore, I removed `seed`.